### PR TITLE
[Feature] New Web UI

### PR DESCRIPTION
**Pull request by ytwp**
## New Web UI
- Developed a new Web UI based on Semi UI, Vite, React, TypeScript, and Echarts frameworks.
- Implemented user interface design and layout using components and styles provided by Semi UI.
- Utilized Vite build tool for fast project development and building.
- Created components and handled interactive logic using React framework.
- Enhanced code safety and maintainability with TypeScript.
- Integrated Echarts charting library for data visualization.

## Screenshot

FireShot Capture 013 - Trino Gateway - localhost

FireShot Capture 006 - Trino Gateway - localhost

FireShot Capture 011 - Trino Gateway - localhost

FireShot Capture 010 - Trino Gateway - localhost

FireShot Capture 009 - Trino Gateway - localhost

FireShot Capture 008 - Trino Gateway - localhost

FireShot Capture 007 - Trino Gateway - localhost

### CODE REVIEW

**Comment by mosabua** [comment]
Wow .. super impressive. Given the large size of the PR we will have to discuss this as a team and figure in detail how we proceed. Couple of initial questions

* Is this a replacement, an upgrade or an additional UI to the existing UI
* are all libraries used transitively Apache licensed or in a similar fashion compatible
* could you change to the new logo?

There is a lot more to discuss in review and in general. Please be patient over the holidays ..  maybe we can arrange a meeting in early January for a core group of project maintainers as next step.

**Comment by andythsu** [comment]
this is super awesome!!

**Comment by ytwp** [comment]
> Wow .. super impressive. Given the large size of the PR we will have to discuss this as a team and figure in detail how we proceed. Couple of initial questions
> 
> * Is this a replacement, an upgrade or an additional UI to the existing UI
> * are all libraries used transitively Apache licensed or in a similar fashion compatible
> * could you change to the new logo?
> 
> There is a lot more to discuss in review and in general. Please be patient over the holidays .. maybe we can arrange a meeting in early January for a core group of project maintainers as next step.

1. This is a replacement to the existing UI.
    
     - But fully compatible with existing UI. Please note that 'privileges' can only be a combination of ADMIN, USER, and API, with '_' used for segmentation.
   
    - The new UI can be launched either independently or together with the 'gateway-ha', depending on the user's preference.

2. All libraries used Apache licensed or MIT licensed.

**Comment by Chaho12** [comment]
> The new UI can be launched either independently or together with the 'gateway-ha', depending on the user's preference.

What do you mean launched independently? you mean that you can launch gateway server withouth web ui?

**Comment by ytwp** [comment]
> > The new UI can be launched either independently or together with the 'gateway-ha', depending on the user's preference.
> 
> What do you mean launched independently? you mean that you can launch gateway server withouth web ui?

This is a front-end and back-end separated architecture, where the pages can be deployed separately on servers like nginx. Just forward the front-end requests to '/webapp/' and it will be relatively more secure.

Alternatively, you can choose to package the front-end static resources into 'gateway' and access them directly through the 'gateway'.

**Comment by mosabua** [comment]
Please rebase, squash commits, and let us know when you are ready for us to review/test. So far this is looking very nice!

**Comment by ytwp** [comment]
> Please rebase, squash commits, and let us know when you are ready for us to review/test. So far this is looking very nice!

I have completed squashing commits. I am ready for review and testing. Thank you for the feedback!

File: gateway-ha/src/main/java/io/trino/gateway/ha/security/LbFormAuthManager.java
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/ha/security/LbFormAuthManager.java:156]: A `List<String>` is preferred here and other places where `String[]` is used

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/request/GlobalPropertyRequest.java
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/request/GlobalPropertyRequest.java:25]: What is the `useSchema` field used for?

**Review by willmostly — COMMENTED** [review]
My comments are limited to the Java portions of this PR. Looks great, just a few items to address.

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:106]: Please use English so that other people can understand easily. Same other places.

File: webapp/.env.production
Comment (Chaho12) [inline webapp/.env.production:3]: Trivial. Add new line please :)

File: webapp/src/api/webapp/dashboard.ts
Comment (Chaho12) [inline webapp/src/api/webapp/dashboard.ts:5]: Here too :)

File: webapp/src/components/layout.module.scss
Comment (Chaho12) [inline webapp/src/components/layout.module.scss:54]: English please :)

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java:98]: Feel free to ignore https://github.com/trinodb/trino-gateway/pull/145. It's just a maintanance PR which should not bock other PRs.

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:24]: Remove unhelpful javadoc. Note that we don't leave `@author`

File: webapp/src/store/access.ts
Comment (Chaho12) [inline webapp/src/store/access.ts:89]: typo? merge?

File: webapp/src/locales/en_US.ts
Comment (Chaho12) [inline webapp/src/locales/en_US.ts:13]: QPH = Query Per Hour? Then is it average too?

File: webapp/src/locales/en_US.ts
Comment (Chaho12) [inline webapp/src/locales/en_US.ts:45]: What do you mean by generating a new account?

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/request/GlobalPropertyRequest.java
Comment (ytwp) [reply gateway-ha/src/main/java/io/trino/gateway/ha/domain/request/GlobalPropertyRequest.java:25]: Please refer to the concept of "useScheme" in the following link: https://github.com/trinodb/trino-gateway/blob/main/docs/resource-groups-api.md.

File: webapp/src/locales/en_US.ts
Comment (ytwp) [reply webapp/src/locales/en_US.ts:13]: It is not an average value because the configuration could be "queryHistoryHoursRetention:1", so there is no design to display average QPH.

File: webapp/src/locales/en_US.ts
Comment (ytwp) [reply webapp/src/locales/en_US.ts:45]: Sorry, I copied it and forgot to delete it

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java
Comment (ytwp) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java:40]: I remember there were issues with Lombok before, but now it's possible to restore to 'boolean' and 'isActive'.

File: webapp/src/locales/zh_CN.ts
Comment (ebyhr) [inline webapp/src/locales/zh_CN.ts:1]: I would recommend handling Chinese support in a follow-up.

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/TableData.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/TableData.java:33]: The javadoc doesn't explain anything. You can remove other javadoc which aren't helpful.

File: webapp/src/locales/zh_CN.ts
Comment (ytwp) [reply webapp/src/locales/zh_CN.ts:1]: Okay

**Comment by andythsu** [comment]
image
The format for this text is different from the rest

The rest of the font has `font-family: Inter,-apple-system,BlinkMacSystemFont,Segoe UI,PingFang SC,Hiragino Sans GB,Microsoft YaHei,Helvetica Neue,Helvetica,Arial,sans-serif;`

But this one is `font-family: PingFang SC;`

**Comment by andythsu** [comment]
Any idea why I keep getting this error when I run this app locally?

image

**Comment by andythsu** [comment]
> > Wow .. super impressive. Given the large size of the PR we will have to discuss this as a team and figure in detail how we proceed. Couple of initial questions
> > 
> > * Is this a replacement, an upgrade or an additional UI to the existing UI
> > * are all libraries used transitively Apache licensed or in a similar fashion compatible
> > * could you change to the new logo?
> > 
> > There is a lot more to discuss in review and in general. Please be patient over the holidays .. maybe we can arrange a meeting in early January for a core group of project maintainers as next step.
> 
> 1. This is a replacement to the existing UI.
>    
>    * But fully compatible with existing UI. Please note that 'privileges' can only be a combination of ADMIN, USER, and API, with '_' used for segmentation.
>    * The new UI can be launched either independently or together with the 'gateway-ha', depending on the user's preference.
> 2. All libraries used Apache licensed or MIT licensed.

Please fix `security.md` doc to include `ADMIN_USER_API`

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java
Comment (andythsu) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java:128]: I had to change it to
```
select FLOOR(created / 1000 / 60) AS minute ...
```
in order for this to work.

**Review by andythsu — CHANGES_REQUESTED** [review]

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java
Comment (andythsu) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java:138]: image

this line returns another error because `model.get("minute")` returns ` 2.8435486E7` that can't be parsed to integer

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java:138]: can be fixed with ` long minute = ((Double) model.get("minute")).longValue();`

**Review by andythsu — CHANGES_REQUESTED** [review]

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java
Comment (ytwp) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java:138]: image

Our results look different. I think different types of databases may return different results. I used MySQL. What type of database are you?

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java:138]: I thought so too. I'm using PostgresSQL

Comment (ytwp) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java]: I fixed this issue

File: docs/security.md
Comment (andythsu) [inline docs/security.md:48]: where/how is this property used?

File: webapp/src/components/login.module.scss
Comment (andythsu) [inline webapp/src/components/login.module.scss]: i think we can just get rid of `.text2{}` and `.text1{}` because its body is empty

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java
Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaQueryHistoryManager.java]: great! Will keep testing it

File: webapp/src/components/cluster.tsx
Comment (andythsu) [inline webapp/src/components/cluster.tsx:184]: is it possible to get rid of these `any` types?

File: webapp/src/components/error.tsx
Comment (andythsu) [inline webapp/src/components/error.tsx:13]: I think we should change this to functional component to stay consistent with the rest of the components

File: webapp/src/components/history.module.scss
Comment (andythsu) [inline webapp/src/components/history.module.scss:1]: extra line here

File: webapp/src/components/resource-group.module.scss
Comment (andythsu) [inline webapp/src/components/resource-group.module.scss:1]: extra line here

File: webapp/src/router.tsx
Comment (andythsu) [inline webapp/src/router.tsx:89]: missing function return type

File: webapp/src/locales/index.ts
Comment (andythsu) [inline webapp/src/locales/index.ts:56]: missing function return type

File: webapp/src/locales/index.ts
Comment (andythsu) [inline webapp/src/locales/index.ts:40]: missing function return type

File: docs/security.md
Comment (ytwp) [reply docs/security.md:48]: If the web UI is started independently, you can specify the redirect path. 

By default, this configuration does not need to be specified.

File: webapp/src/components/error.tsx
Comment (ytwp) [reply webapp/src/components/error.tsx:13]: Because function components don't provide 'componentDidCatch', so here I am using class components directly without introducing any third-party libraries to implement it.

File: webapp/src/components/cluster.tsx
Comment (ytwp) [reply webapp/src/components/cluster.tsx]: 👌 Okay, change it to a specific type.

**Review by Chaho12 — COMMENTED** [review]
For those who are testing, please note that
- UI requires AUTHORIZATION, so `docs/quickstart-config.yaml` doesnt' work properly unless presetUsers and authentication is set
  - Can this be optional?

I've got some suggestions for improvements in UI
- In login screen, can you fix so that enter works to login? kinda bothersome to click on sign in button everytime
image

- In history tab, how about adding `<all>` or sth placeholder for `RoutedTo` and `QueryId`  which means all (just for the ui part like if none is selected, it means all in history) 
image
- add hyperlink at `Backends` to link cluster in dashboard for convenience?
image
- add a button next to QPH/Avg. QPM/Avg. QPS to explain how each value is calculated? 
image

File: webapp/src/locales/en_US.ts
Comment (Chaho12) [inline webapp/src/locales/en_US.ts:14]:
```suggestion
    QueryDistribution: "Query Distribution in last hour.",
```

**Comment by vishalya** [comment]
+1 for making the authentication making optional.

**Comment by ytwp** [comment]
> For those who are testing, please note that
> 
> * UI requires AUTHORIZATION, so `docs/quickstart-config.yaml` doesnt' work properly unless presetUsers and authentication is set
>   
>   * Can this be optional?
> 
> I've got some suggestions for improvements in UI
> 
> * In login screen, can you fix so that enter works to login? kinda bothersome to click on sign in button everytime
>   image
> * In history tab, how about adding `<all>` or sth placeholder for `RoutedTo` and `QueryId`  which means all (just for the ui part like if none is selected, it means all in history)
>   image
> * add hyperlink at `Backends` to link cluster in dashboard for convenience?
>   image
> * add a button next to QPH/Avg. QPM/Avg. QPS to explain how each value is calculated?
>   image

👍 Great suggestion, currently under adjustment.

**Comment by ytwp** [comment]
Hi, @Chaho12 @vishalya 

If authentication is made optional, what permissions should be granted?
My suggestion is to add a default administrator account in the default configuration file.

**Comment by mosabua** [comment]
How about using the same approach as in the Trino UI .. username is required but can be random and you get full access.

**Comment by mosabua** [comment]
Fyi @ytwp ... in our dev sync earlier today we discussed the PR. We decided that we want to go ahead towards merge. A next step would be for you to rebase so its ready and then we can review. We will just get it to a good enough state for merge and plan on iterating on it as necessary together as well. The one thing we need to make sure is that we remove the requirement for authentication similar to how it is optional in Trino.

Ping us on slack if you want to chat or let us know here.

**Review by Chaho12 — APPROVED** [review]
LGTM 👍

File: gateway-ha/pom.xml
Comment (ebyhr) [inline gateway-ha/pom.xml:240]: Avoid introducing new utility dependency. We already have guava.

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:24]: Avoid using abbreviations. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-abbreviations Same for others.

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:176]: Remove unused methods in this class.

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/TableData.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/domain/TableData.java:24]: Add jackson annotations. Same for others.

File: gateway-ha/src/main/java/io/trino/gateway/ha/module/HaGatewayProviderModule.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/module/HaGatewayProviderModule.java:109]: Unwrap.

File: gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java:19]: final

File: gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java:18]: Remove unhelpful comment. Same for others.

File: gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java:24]: nit
```suggestion
    private PageUtil() {}
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/util/PageUtil.java:21]: Uppercase a constant name

File: webapp/src/locales/index.ts
Comment (ebyhr) [inline webapp/src/locales/index.ts:19]: Remove a language button from the header. It's unused at this moment.

**Comment by ebyhr** [comment]
> [Feature] New Web UI

Please update the commit title as "Modernize Web UI" or something. Also, please add the commit body to describe the changes. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

File: gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java
Comment (ytwp) [reply gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:176]: I suggest keeping the unused methods in this class so they can be easily used in the future.👀👀

Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:176]: Disagreed. We can add these methods easily when it's exactly needed. Please remove them.

Comment (mosabua) [reply gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:176]: Yeah .. we should not carry around any stuff that is not actually used.

Comment (ytwp) [reply gateway-ha/src/main/java/io/trino/gateway/ha/domain/R.java:176]: Ok

**Review by mosabua — APPROVED** [review]
I ran a bunch of tests with this and looked over the whole codebase changes. This is a very significant improvement from the legacy UI. 

It creates a base for further improvements that we can work on in follow up PRs where necessary, but the current state is already good for merge.

We will do further testing after merge, also with other features merged, and in the build up to the next release.

**Comment by mosabua** [comment]
I will update the commit message as part of the merge.

**Pull request merged**
